Convert df to pyspark DataFrame if it is pandas before writing

dbeatty10 opened the pull request:
resolves #468

### Description

Copies the solution by @chamini2 in https://github.com/dbt-labs/dbt-bigquery/pull/301

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-spark/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-spark/blob/main/CONTRIBUTING.md#Adding-CHANGELOG-Entry)

dbeatty10 commented:
### Overview
I manually verified that the following didn't work before for the dbt-databrick adapter (which inherits from dbt-spark). Confirmed that it works using @chamini2's fix 👍

<details>
  <summary>Code example</summary>

```python
import pandas as pd

def model(dbt, session):
    dbt.config(
        materialized="table",
        packages=["pandas"]
    )

    df = pd.DataFrame(
        {'City': ['Buenos Aires', 'Brasilia', 'Santiago', 'Bogota', 'Caracas'],
        'Country': ['Argentina', 'Brazil', 'Chile', 'Colombia', 'Venezuela'],
        'Latitude': [-34.58, -15.78, -33.45, 4.60, 10.48],
        'Longitude': [-58.66, -47.91, -70.66, -74.08, -66.86]}
        )

    return df
```
</details>

### Details
dbt-databricks has [it's own implementation of `py_write_table`](https://github.com/databricks/dbt-databricks/blob/fe61d6fa539c8a8eb3d43ca6565fd4a16cbcb411/dbt/include/databricks/macros/materializations/table.sql#L40-L46) which overrides the implementation provided by dbt-spark.

⚠️ As a result, Databricks users won't have support for Pandas DataFrames until one of the following happens:
1. dbt-databricks deletes the custom implementation of `py_write_table`
1. This implementation is copied into dbt-databricks also
1. Some other update is added to dbt-databricks that enables Pandas DataFrames

We want to provide an equivalent implementation here so that dbt-databricks has the option to fully drop their implementation of `py_write_table`. So this PR adopts two changes from dbt-databricks's version of `py_write_table`:
1. `.option("overwriteSchema", "true")`
1. `# --- Autogenerated dbt materialization code. --- #`

### Follow-up
I will open an issue in dbt-databricks for this.

b-per commented on dbt/include/spark/macros/materializations/table.sql:
```diff
@@ -41,7 +41,17 @@
+import importlib.util
+package_name = 'pandas'
+if importlib.util.find_spec(package_name):
+    import pandas
```
Is this line required?

From what I see, in databricks we might usually want to load `pyspark.pandas` rather than `pandas` (and I don't think it is required here)

dbeatty10 replied:
@b-per I believe it _is_ required in order to do test the type: `isinstance(df, pandas.core.frame.DataFrame)`. I tried commenting out the single line with the import, and it did not work.

However, digging into your question did uncover something else...

### Goal
For the `py_write_table()` macro, we want the return type to be `pyspark.sql.dataframe.DataFrame`.

### Potential input types

There are three different data types we expect <sup>[1](#internal_references)</sup> it to be:
1. `pyspark.sql.dataframe.DataFrame` (PySpark DataFrame)
1. `pandas.core.frame.DataFrame` (Pandas DataFrame)
1. `pyspark.pandas.frame.DataFrame` (Pandas-on-Spark DataFrame)

Are we handling each of the three cases?
1. ✅ For the first case, it is already a `pyspark.sql.dataframe.DataFrame`, so no conversion necessary.
1. ✅ For the second case, this PR infers if it is a `pandas.core.frame.DataFrame` and converts it using `spark.createDataFrame()` if so.
1. ❌ However, the third case isn't handled yet! We can handle it by calling `to_spark()` on it <sup>[2](#internal_references)</sup>.

I'll update this PR to include the third case.

---

<a name="internal_references"></a>

1. We could proactively raise an exception if it falls outside of these expected types (rather than sending it off to the database where it will fall on its face and emit an unhelpful error message).
1. Thank you to @Adricarpin for this handy resource: https://towardsdatascience.com/run-pandas-as-fast-as-spark-f5eefe780c45#64b7

b-per replied:
Yep, my bad, I thought `isinstance` was comparing with a string and not an object type.

dbeatty10 replied:
@b-per Very glad you asked, because now we have a more robust implementation as a result!

b-per reviewed (changes requested):
I think we can slightly update the code (but I could be wrong as well)

chamini2 commented on dbt/include/spark/macros/materializations/table.sql:
```diff
@@ -41,7 +41,20 @@
+    # convert to pyspark.sql.dataframe.DataFrame
+    if isinstance(df, pandas.core.frame.DataFrame):
+      df = spark.createDataFrame(df)
+    elif isinstance(df, pyspark.pandas.frame.DataFrame):
```
I think this can happen **outside** the `if importlib.util.find_spec(package_name):`. Since it would run only for systems with pandas installed.

dbeatty10 replied:
Thanks @chamini2 -- I manually tested that suggestion and pushed a commit 👍

dbeatty10 commented:
@ChenyuLInx see below for the Python model files that I used for manual testing. [Example data](https://geopandas.org/en/v0.11.0/gallery/create_geopandas_from_pandas.html#From-longitudes-and-latitudes) is from the GeoPandas documentation.

### Pandas DataFrame
```python
# models/pandas_df.py

import pandas as pd


def model(dbt, session):
    dbt.config(
        materialized="table",
        packages=["pandas"]
    )

    df = pd.DataFrame(
        {'City': ['Buenos Aires', 'Brasilia', 'Santiago', 'Bogota', 'Caracas'],
        'Country': ['Argentina', 'Brazil', 'Chile', 'Colombia', 'Venezuela'],
        'Latitude': [-34.58, -15.78, -33.45, 4.60, 10.48],
        'Longitude': [-58.66, -47.91, -70.66, -74.08, -66.86]}
        )

    return df
```

### PySpark DataFrame
```python
# models/pyspark_df.py

def model(dbt, session):
    dbt.config(
        materialized="table",
    )

    df = spark.createDataFrame(
        [
            ("Buenos Aires", "Argentina", -34.58, -58.66),
            ("Brasilia", "Brazil", -15.78, -47.91),
            ("Santiago", "Chile", -33.45, -70.66),
            ("Bogota", "Colombia", 4.60, -74.08),
            ("Caracas", "Venezuela", 10.48, -66.86),
        ],
        ["City", "Country", "Latitude", "Longitude"]
    )

    return df
```

### Pandas-on-Spark DataFrame
```python
# models/pandas_on_spark_df.py

import pyspark.pandas as ps


def model(dbt, session):
    dbt.config(
        materialized="table",
    )

    df = ps.DataFrame(
        {'City': ['Buenos Aires', 'Brasilia', 'Santiago', 'Bogota', 'Caracas'],
        'Country': ['Argentina', 'Brazil', 'Chile', 'Colombia', 'Venezuela'],
        'Latitude': [-34.58, -15.78, -33.45, 4.60, 10.48],
        'Longitude': [-58.66, -47.91, -70.66, -74.08, -66.86]}
        )

    return df
```

### Pandas-on-Spark DataFrame to Pandas DataFrame
```python
# models/pandas_on_spark_df_to_pandas.py

import pyspark.pandas as ps


def model(dbt, session):
    dbt.config(
        materialized="table",
    )

    df = ps.DataFrame(
        {'City': ['Buenos Aires', 'Brasilia', 'Santiago', 'Bogota', 'Caracas'],
        'Country': ['Argentina', 'Brazil', 'Chile', 'Colombia', 'Venezuela'],
        'Latitude': [-34.58, -15.78, -33.45, 4.60, 10.48],
        'Longitude': [-58.66, -47.91, -70.66, -74.08, -66.86]}
        )
    pdf = df.to_pandas()

    return pdf
```

chamini2 commented on dbt/include/spark/macros/materializations/table.sql:
```diff
@@ -41,7 +41,21 @@
+    import pandas
+
+# convert to pyspark.sql.dataframe.DataFrame
+if isinstance(df, pandas.core.frame.DataFrame):
```
But this has to happen inside of the `importlib.util.find_spec(package_name):` if right? If not, you will be accessing a module `pandas` that is not really imported if it was not present.

chamini2 commented on dbt/include/spark/macros/materializations/table.sql:
```diff
@@ -41,7 +41,21 @@
-df.write.mode("overwrite").format("delta").saveAsTable("{{ target_relation }}")
+
+# make sure pandas exists
+import importlib.util
+import pyspark.pandas
+package_name = 'pandas'
+if importlib.util.find_spec(package_name):
+    import pandas
+
+# convert to pyspark.sql.dataframe.DataFrame
+if isinstance(df, pandas.core.frame.DataFrame):
+  df = spark.createDataFrame(df)
+elif isinstance(df, pyspark.pandas.frame.DataFrame):
+  df = df.to_spark()
```
```suggestion
# convert to pyspark.sql.dataframe.DataFrame
import importlib.util
if importlib.util.find_spec('pandas'):
    import pandas
    if isinstance(df, pandas.core.frame.DataFrame):
        df = spark.createDataFrame(df)

import pyspark.pandas
if isinstance(df, pyspark.pandas.frame.DataFrame):
  df = df.to_spark()
```

chamini2 replied:
This is how I would add the pyspark.pandas check

dbeatty10 commented:
Incorporated feedback from @chamini2, @ueshin, and @ChenyuLInx.

Added these features:
- Also check the availability of `pyspark.pandas` since it was introduced in Spark 3.2 and Databricks still supports DBR 9.1 that is based on Spark 3.1.2
- Raise an error exception if unable to convert to a Spark DataFrame (`pyspark.sql.dataframe.DataFrame`)

The net effect is:
- No need to convert Spark DataFrames -- already the type we want
- Convert pandas DataFrames to Spark DataFrame
- Convert pandas-on-Spark DataFrames to Spark DataFrame
- Raise an exception for all other cases

chamini2 reviewed (approved):
Looks great!

jtcohen6 commented on dbt/include/spark/macros/materializations/table.sql:
```diff
@@ -41,7 +41,40 @@
+# since they know how to convert pandas DataFrames better than `spark.createDataFrame(df)`
+# and converting from pandas-on-Spark to Spark DataFrame has no overhead
+if pyspark_available and pandas_available and isinstance(df, pandas.core.frame.DataFrame):
+  df = pyspark.pandas.frame.DataFrame(df)
```
At the risk of making this even more complex than it needs to be — I believe `pyspark.pandas` was introduced in Spark v3.2, and won't be available in earlier versions. (The same functionality was available via the `koalas` package, which was the old codename for pandas-on-PySpark)

dbeatty10 replied:
I _think_ what will happen currently if the input is a pandas-on-Spark DataFrame but `pyspark.pandas` is not available:
-   `msg = f"{type(df)} is not a supported type for dbt Python materialization"`

What I believe will happen if the input is a pandas DataFrame but `pyspark.pandas` is not available:
-   `df = spark.createDataFrame(df)`

We can add in an attempt to `import databricks.koalas` so we are covering as many bases as possible. If we go that route, is there an environment we could test it out on?

jtcohen6 replied:
> If we go that route, is there an environment we could test it out on?

We could spin up a Databricks cluster running an older Spark version (v3.1). This is also what will be running inside Dataproc — the latest Apache Spark release it supports is v3.1.
